Add tags in task results details. Add tags in as tooltip in task results

JayV3e opened the pull request:
This fix issue #256.

In task results, I added a tag button with the number of tags.
On mouse over, there is a tooltip that shows the tags.

Let me know if I should stick to the previous pattern of mouse over the task name instead of an additional span.

<img width="1238" alt="Screen Shot 2021-06-26 at 5 12 08 PM" src="https://user-images.githubusercontent.com/12644084/123525905-20e72f80-d6a2-11eb-8d0a-a24d77a2c4ac.png">

In task result, I added the Tags if there is any. Otherwise, the Tags: won't show up.

With tags : 
<img width="549" alt="Screen Shot 2021-06-26 at 5 12 21 PM" src="https://user-images.githubusercontent.com/12644084/123525958-96530000-d6a2-11eb-8683-1e99ff12402e.png">

Without tags : 
<img width="525" alt="Screen Shot 2021-06-26 at 5 12 31 PM" src="https://user-images.githubusercontent.com/12644084/123525965-9f43d180-d6a2-11eb-8949-21f9df9d7445.png">

I added the <script> and <style> directly in the playbook_results.html template.
I'm not sure if that's a good practice or not. 

Waiting for feedback on the tag span, where to put the <script> and <style> or any additional feedback. 

This is my first pull request on an open source project, let me know if I forgot anything relevant.

dmsimard commented on ara/ui/templates/partials/playbook_results.html:
```diff
@@ -149,6 +149,11 @@ <h4>
                       {{ result.task.name }}
                     {% endif %}
+                    {% if result.task.tags %}
+                    <span class="badge badge-pill badge-warning" class="btn btn-secondary" data-toggle="tooltip"  data-placement="right" data-html="true" title="{% include 'partials/tags_list.html' with tags=result.task.tags %}">
```
I'm OK with this approach but I would tend to use another color here because it overlaps with what we use to highlight the changed status:
![Screenshot from 2021-06-26 21-05-09](https://user-images.githubusercontent.com/1291204/123529836-5d2a8800-d6c2-11eb-9fbb-e45467315056.png)

Looking at the [docs](https://getbootstrap.com/docs/4.0/components/badge/#contextual-variations), maybe ``badge-secondary`` instead of ``badge-warning`` ?

dmsimard commented:
It looks good to me although I wonder if it's the right place or column for the badge pill because tasks have variable length names so the badge can display a bit all over the place. I wonder what it would look like in the status column ?

We would also need to add the tags in the same place for the host results table: https://github.com/ansible-community/ara/blob/master/ara/ui/templates/partials/host_results.html#L160

dmsimard commented:
Also: ignore the failing container job, it's unrelated to your patch and I need to fix it.

JayV3e commented:
I changed color and column.

By default, it looks awful : 
<img width="1255" alt="Screen Shot 2021-06-27 at 4 34 09 PM" src="https://user-images.githubusercontent.com/12644084/123558725-ca432980-d765-11eb-99fd-c3cdcc272c92.png">

We could either add the label with 0 tags while removing the empty tooltip : 
<img width="1248" alt="Screen Shot 2021-06-27 at 4 34 34 PM" src="https://user-images.githubusercontent.com/12644084/123558737-db8c3600-d765-11eb-8016-881e5665bcdc.png">

Or simply add "invisible" as a class for the span : 
<img width="1270" alt="Screen Shot 2021-06-27 at 4 35 16 PM" src="https://user-images.githubusercontent.com/12644084/123558742-e3e47100-d765-11eb-85ed-4ef0e70db14e.png">

I also added the same label to the host result : 
<img width="1111" alt="Screen Shot 2021-06-27 at 4 43 15 PM" src="https://user-images.githubusercontent.com/12644084/123558888-cebc1200-d766-11eb-80fc-f589ff6710d0.png">


Initially I thought invisible would've been the best solution but I kinda like the 0 tags as well.

Let me know whichever you prefer.

flowerysong commented:
While this should definitely be included in the detailed view, I don't really feel this is important enough information to surface it so prominently.

JayV3e commented:
I agree. My example only has 3 tags and it already takes quite some space. I don't know how many tags people use on average but this could easily get out of hands.

Here's 3 suggestions : 
- Make a list of the tags

Ex : 
Tags: tag1,tag2,tag3,tag4

- Keep it as it is but make it collapsed by default and expendable.

- Use the same pattern as in the playbook result: Have a label to show to number of tags and have the list appear on mouse over.

My vote goes for the last one.

Let me know if you have another suggestion.

dmsimard commented:
> While this should definitely be included in the detailed view, I don't really feel this is important enough information to surface it so prominently.

Thanks for the feedback @flowerysong !
I agree to some extent which is in part why tags had not yet made their way to the results table yet :)

The task tags are included by default "for free" in the API response, we're just not displaying them anywhere right now. I figure it would be useful to at least do something with them.

In ara 0.x, it used to be a tooltip over the task name, like this: https://github.com/ansible-community/ara/blob/stable/0.x/ara/templates/ajax/task_name.html
![Screenshot from 2021-06-30 15-44-29](https://user-images.githubusercontent.com/1291204/124021926-256d5a00-d9ba-11eb-87ae-d78d19d76910.png)

That implementation wasn't perfect either because you only knew if the task had tags at all if you put your mouse over the task name so it wasn't obvious from a UX perspective which tasks had tags or not.

Maybe we could go with that approach instead ? There is no debate to be had about the detailed view, we all agree it would be a nice addition.

dmsimard commented:
Hi @JayV3e o/

Sorry about the delay -- I've taken a bit of time to think about this and asked different users what they thought.

Given the choice, not everyone prefers the same one (UI/UX is hard!) but there's a small majority that would go with the third option you suggested:
![image](https://user-images.githubusercontent.com/12644084/123558888-cebc1200-d766-11eb-80fc-f589ff6710d0.png)

If that's okay with you, we'll go with that for now and I'm happy to revisit later if we find a better implementation.

dmsimard commented:
This isn't a blocker for releasing 1.5.7 which I would like to tag, it can go in the next release.

Feel free to let me know when you see my comments @JayV3e o/

dmsimard commented on ara/ui/templates/partials/host_results.html:
```diff
@@ -145,6 +145,13 @@
                     {% else %}
                       {% include "partials/result_status_icon.html" with status=result.status %}
                     {% endif %}
+                    {% if result.task.tags|length == 0 %}
```
nitpick, would rather have the version with the space like the one you wrote below for consistency

```suggestion
                    {% if result.task.tags | length == 0 %}
```

dmsimard commented on ara/ui/templates/partials/playbook_results.html:
```diff
@@ -127,6 +127,13 @@ <h4>
                     {% else %}
                       {% include "partials/result_status_icon.html" with status=result.status %}
                     {% endif %}
+                    {% if result.task.tags|length == 0 %}
```
ditto
```suggestion
                    {% if result.task.tags | length == 0 %}
```

dmsimard commented on ara/ui/templates/partials/tags_list.html:
```diff
@@ -0,0 +1,5 @@
+<ul>
+    {% for tag in tags %}
+    <li> {{tag}} </li>
```
```suggestion
    <li> {{ tag }} </li>
```

JayV3e commented:
Hello,

I applied all suggestions and changed the tags formatting in task results to use bullet points.

Hope I did it in time for the release !

dmsimard commented:
The patch would have been fine but the CI is now reporting an issue with the PR when running ``ara-manage generate`` to generate a static version of the report:
```
[ara] Using settings file: /home/zuul/.ara-tests/server/settings.yaml
[ara] Generating static files for 8 playbooks at /home/zuul/.ara-tests/server/static...
Traceback (most recent call last):
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.9/site-packages/django/template/smartif.py", line 175, in translate_token
    op = OPERATORS[token]
KeyError: '|'

During handling of the above exception, another exception occurred:
# ...
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.9/site-packages/django/template/base.py", line 662, in __init__
    raise TemplateSyntaxError("Could not parse the remainder: '%s' "
django.template.exceptions.TemplateSyntaxError: Could not parse the remainder: '|' from '|'
```

That's odd because it was green until the last commit:
![Screenshot from 2021-08-02 14-28-26](https://user-images.githubusercontent.com/1291204/127907187-7c77ecff-623b-4cea-ad56-2879d7d4f5e7.png)

I will need to investigate after the release. Sorry !

dmsimard commented on ara/ui/templates/partials/host_results.html:
```diff
@@ -145,6 +145,13 @@
                     {% else %}
                       {% include "partials/result_status_icon.html" with status=result.status %}
                     {% endif %}
+                    {% if result.task.tags | length == 0 %}
+                      <span class="badge badge-pill badge-secondary invisible" class="btn btn-secondary">
+                    {% else %}
+                      <span class="badge badge-pill badge-secondary" class="btn btn-secondary" data-toggle="tooltip"  data-placement="right" data-html="true" title="{% include 'partials/tags_list.html' with tags=result.task.tags %}">
+                    {% endif %}
+                      {{ result.task.tags | length }} tags
+                      </span>
```
I was able to reproduce the issue I mentioned in my [last comment](https://github.com/ansible-community/ara/pull/281#issuecomment-891240781) locally and it's indeed tripping over the spaces before and after the pipe -- if there are no spaces it works. I can reproduce the issue with the local server running, no need to generate a static report.

This is surprising to me because there's plenty of places where we pipe templatetags with spaces around the pipe (like [here](https://github.com/ansible-community/ara/blob/7ce19a37c0c42cc0ad2b36b3ebed19bed99ff0df/ara/ui/templates/partials/playbook_results.html#L161-L162)) so I'm mildly confused at it -- it may have to do with being inside an if block ? I'll have learned something today.

Anyway, I suggest switching the condition around so we don't need to test for the length (an empty list is false):

```suggestion
                    {% if result.task.tags %}
                      <span class="badge badge-pill badge-secondary" class="btn btn-secondary" data-toggle="tooltip"  data-placement="right" data-html="true" title="{% include 'partials/tags_list.html' with tags=result.task.tags %}">
                    {% else %}
                      <span class="badge badge-pill badge-secondary invisible" class="btn btn-secondary">
                    {% endif %}
                      {{ result.task.tags | length }} tag{{ result.task.tags | pluralize }}
                      </span>
```

I've also added a ``| pluralize`` filter that automatically adds an "s" if there is >1 tag.

dmsimard commented on ara/ui/templates/partials/playbook_results.html:
```diff
@@ -127,6 +127,13 @@ <h4>
                     {% else %}
                       {% include "partials/result_status_icon.html" with status=result.status %}
                     {% endif %}
+                    {% if result.task.tags | length == 0 %}
+                      <span class="badge badge-pill badge-secondary invisible" class="btn btn-secondary">
+                    {% else %}
+                      <span class="badge badge-pill badge-secondary" class="btn btn-secondary" data-toggle="tooltip"  data-placement="right" data-html="true" title="{% include 'partials/tags_list.html' with tags=result.task.tags %}">
+                    {% endif %}
+                      {{ result.task.tags | length }} tags
+                      </span>
```
```suggestion
                    {% if result.task.tags %}
                      <span class="badge badge-pill badge-secondary" class="btn btn-secondary" data-toggle="tooltip"  data-placement="right" data-html="true" title="{% include 'partials/tags_list.html' with tags=result.task.tags %}">
                    {% else %}
                      <span class="badge badge-pill badge-secondary invisible" class="btn btn-secondary">
                    {% endif %}
                      {{ result.task.tags | length }} tags
                      </span>
```

dmsimard commented on ara/ui/templates/partials/playbook_results.html:
```diff
@@ -174,3 +181,14 @@ <h3>No results found matching your query: try <a href="{% url 'ui:playbook' play
+  .tooltip {
+    pointer-events: none;
+  }
+</style>
```
Could you add the ``<script>`` to host_results.html too ? The tooltip doesn't work there because it's missing.

For the style block, could you move it to [ara.css](https://github.com/ansible-community/ara/blob/master/ara/ui/static/css/ara.css) instead ?

Thanks !

JayV3e commented:
I move the <style> to ara.css

I didn't know about pluralize, this is pretty sweet. Thanks for the tip!

All suggestions have been applied.

dmsimard commented:
I should have merged this sooner, I'll tweak it a little bit based on feedback but it'll go in 1.5.8. Thanks !

dmsimard commented:
Tweaks in this PR: https://github.com/ansible-community/ara/pull/375
